kernel_modules/kmod_meltdown/meltdown_module.c

```diff
@@ -7,96 +7,95 @@
  * SPDX-License-Identifier: BSD-3-Clause OR GPL-2.0
  */
 
+#include <linux/debugfs.h>
 #include <linux/module.h>
-#include <linux/proc_fs.h>
-#include <linux/seq_file.h>
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_AUTHOR("Google");
 MODULE_DESCRIPTION("");
-MODULE_VERSION("0.1");
-
-// Keeps a buffer in kernel memory and makes its address available at:
-//   /proc/safeside_meltdown/address
-// and its length available at:
-//   /proc/safeside_meltdown/length
-// which are both only accessible to root.
-
-// Secret data stored in the kernel memory whose content is never directly
-// leaked through sysfs.
-const char *private_data = "It's a s3kr3t!!!";
-
-// Directory record. Must be available on unloading the module.
-struct proc_dir_entry *safeside_meltdown;
-
-// Print the address of `private_data` into the provided buffer.
-static int address_show(struct seq_file *file, void *v) {
-  seq_printf(file, "%px\n", private_data);
-  return 0;
-}
+MODULE_VERSION("0.2");
+
+// This module provides the kernel side of a test for Meltdown.
+//
+// It holds a buffer of "secret" data in kernel memory that is never made
+// (legitimately) available to userspace. Through debugfs, userspace can find
+// the address and length of that buffer and can force it to be read into
+// cache.
+//
+// Assuming debugfs is mounted in the usual place, the files are at:
+//   /sys/kernel/debug/safeside_meltdown/
+// And they are:
+//   secret_data_address
+//     virtual address of secret data, returned as "0x123"
+//   secret_data_length
+//     length of secret data in bytes, returned as "123"
+//   secret_data_in_cache
+//     when opened and read, forces secret data to be read into cache.
+//
+// Access is restricted to root to avoid opening any new attack surface, e.g.
+// leaking kernel pointers.
+
+// The notionally secret data.
+static const char secret_data[] = "It's a s3kr3t!!!";;
+
+// Values published through debugfs.
+static u64 secret_data_address = (u64)secret_data;
+static u32 secret_data_length = sizeof(secret_data);
+
+// The dentry of our debugfs folder.
+static struct dentry *debugfs_dir = NULL;
+
+// Getter that reads secret_data into cache.
+int secret_data_in_cache_get(void* data, u64* out) {
+  int i;
+  for (i = 0; i < secret_data_length; ++i) {
+    *(volatile char *)&secret_data[i];
+  }
 
-// Print the length of `private_data` into the sequential file.
-// At the same time loads `private_data` to cache by accessing it.
-static int length_show(struct seq_file *file, void *v) {
-  seq_printf(file, "%d\n", (int) strlen(private_data));
+  *out = 1;  // As in, "true", the secret data _is_ now in cache.
```

Comment:
The meaning of out should be documented in the function comment instead of here IMO.

Reply:
Done.